packages/filesystem/src/browser/file-resource.ts

```diff
@@ -92,11 +92,30 @@ export class FileResource implements Resource {
             console.error(e);
         }
         this.updateSavingContentChanges();
-        this.toDispose.push(this.fileService.onDidChangeFileSystemProviderCapabilities(e => {
+        this.toDispose.push(this.fileService.onDidChangeFileSystemProviderCapabilities(async e => {
             if (e.scheme === this.uri.scheme) {
-                this.updateSavingContentChanges();
+                this.updateReadOnly();
             }
         }));
+        this.fileService.onDidChangeFileSystemProviderReadOnlyMessage(async e => {
+            if (e.scheme === this.uri.scheme) {
+                this.updateReadOnly();
+            }
+        });
+    }
+
+    protected async updateReadOnly(): Promise<void> {
+        const oldReadOnly = this.options.readOnly;
+        const readOnlyMessage = this.fileService.readOnlyMessage(this.uri);
+        if (readOnlyMessage) {
+            this.options.readOnly = readOnlyMessage;
+        } else {
+            this.options.readOnly = await this.fileService.isReadOnly(this.uri) || this.fileService.hasCapability(this.uri, FileSystemProviderCapabilities.Readonly);
```

Comment:
Suggestion: I'm not sure whether we should actually call the `fileService.isReadOnly()` method here. While this has the benefit of reflecting the current readonly state of the file we're reading (and potentially making it readonly/writable based on the actual file system permissions of the file), it is only updated when the fs capability changes. This feels pretty inconsequential, as the change isn't immediately reflected in the editor - only after we change the capabilities (which is pretty non-trivial from a user perspective). I would remove the `isReadOnly` call here IMO.

Unrelated to that, I have another suggestion that would improve the need to call `isReadOnly` in certain situations:
```suggestion
            this.options.readOnly = this.fileService.hasCapability(this.uri, FileSystemProviderCapabilities.Readonly) || await this.fileService.isReadOnly(this.uri);
```